Add JSON support in extra_vars

cognifloyd opened the pull request:
The docs[1] say that ansible supports json/yaml in extra_vars. Fixes #6. This adds support for sending arbitrarily complex extra_vars (such as when specifying parameters in a workflow) as json to ansible.

This also changes the concatenation of key=value pairs so that only contiguous key=value pairs are merged into the same extra-vars arg. That way, order is preserved in the rare case where someone overrides a variable in a subsequent extra_vars entry.

[1] http://docs.ansible.com/ansible/playbooks_variables.html#passing-variables-on-the-command-line

cognifloyd committed:
Add JSON support in extra_vars

Ansible supports json/yaml in extra_vars[1]. This commit allows
stackstorm to pass more complex extra_vars parameters (such as
objects/dicts) to ansible as JSON.

One interesting bit to note: The ansible action runners get the
extra_vars parameters as ast literals on the command line. As they
are accessed via ast.literal_eval, sending anything other than a
string caused things to break before (dicts don't have a strip()
method).

Also, this preserves the order of extra_vars so that if a var
overrides a particular file or dict, it can override it. Contiguous
lists of key=value vars are still combined into one --extra_vars
parameter, but if they are separated by @files or dicts, then the
order is preserved.

[1] http://docs.ansible.com/ansible/playbooks_variables.html#passing-variables-on-the-command-line

cognifloyd commented on actions/lib/ansible_base.py:
```diff
@@ -43,10 +43,13 @@ def _parse_extra_vars(self):
                 var_list = [n.strip() for n in var_list]
```
This line will need to change because it assumes that ast.literal_eval will provide strings. However, I gave it a json object / dict.

cognifloyd commented on actions/lib/ansible_base.py:
```diff
@@ -43,10 +43,13 @@ def _parse_extra_vars(self):
                 kwarg_vars = list()
                 for v in var_list:
-                    if "@" in v:
+                    if v.strip().startswith("{"):
```
Pointless, as ast.literal_eval already occurred. It's not a string.

cognifloyd commented on actions/lib/ansible_base.py:
```diff
@@ -43,10 +43,13 @@ def _parse_extra_vars(self):
                 # Separate @file vars from kwarg vars
+                json_vars = list()
                 file_vars = list()
                 kwarg_vars = list()
```
Splitting the vars between file_vars and kwarg_vars like this could have unforseen ordering problems. If you specifically set one var before or after a file, that ordering is lost here.

cognifloyd replied:
So, maybe instead of separate dicts, the var_list should be converted to a list of tuples (type, var)

cognifloyd reviewed (commented):
There are several issues with this first approach and with how it works currently. I think list_vars needs to become a list of two tuples: `(type, var)` where type is one of file, kwarg, or json.

cognifloyd commented:
I squashed my commits, fixed various doc strings, added an entry to CHANGES.md, and bumped the version to 0.5.0. I think this is ready to be merged.

arm4b reviewed (changes requested):
Good feature!
Can you also add a description and example in README.md?

Additionally, this method becomes really complex. Adding some simple tests would be nice.

arm4b commented on actions/command.yaml:
```diff
@@ -53,7 +53,7 @@ parameters:
     type: string
   extra_vars:
-    description: "List of additional variables to pass to ansible. Each variable is represented as \"key=value\" or \"@path/to/file.yaml|json\" [-e]."
+    description: "List of additional variables to pass to ansible. Each variable is represented as \"key=value\" or \"@path/to/file.yaml|json\" or use a yaml dict (\"key: [1, 2, 3]\") to send JSON. [-e]"
```
Just a recommendation: we can replace it with something like `description: '..."key=value"...'` to avoid escaping double quotes which don't look nice.

cognifloyd replied:
:heavy_check_mark: Fixed description strings

arm4b commented:
@chris-downs would be nice if you could take a look too, since you initiated adding advanced functionality for extra vars.

cognifloyd committed:
Clean up action descriptions

cognifloyd committed:
Add example to readme

cognifloyd reviewed (commented):
Fixed descriptions and added an example to the readme.

arm4b commented:
Thanks for adding descriptions! :+1:

We still need some simple unit tests for the `_parse_extra_vars` method. With the increased complexity, we can't afford that luxury of not writing tests anymore.

cognifloyd commented:
I have to learn about unit tests in python. I think there's a doc page about testing packs, so I'll dig into that later.

cognifloyd committed:
Add tests for extra_vars parsing

Also, based on tests, make it actually work.

The extra_vars.yaml fixture-based tests were written based on the master
code base. One test had to change based on my order-preserving change,
and the expected results are commented out to show the difference
between the old output and the new. The json and complex tests ensure
that the new json functionality works as expected.

Signed-off-by: Jacob Floyd <cognifloyd@gmail.com>

cognifloyd commented:
There are now 14 tests for _parse_extra_vars, 7 of which were written to pass with the current master (of those 7, only one test needed to be changed to account for order-preservation of arguments in this version). The other 7 test the json functionality. I did not add tests for any other code paths.

Thanks to @nmaludy, @m4dcoder, and @lakshmi-kannan for helping me figure out how to put these tests together.

arm4b reviewed (approved):
Awesome!

@cognifloyd Additional big thanks for the tests :+1:

cognifloyd committed:
Fix whitespace

Signed-off-by: Jacob Floyd <cognifloyd@gmail.com>

cognifloyd commented:
I saw a simple whitespace issue in the readme, so I fixed it. This is ready to merge imo.
